src/docfx/validation/JsonSchemaValidatorExtension.cs

```diff
@@ -46,5 +56,202 @@ public bool IsEnable(FilePath filePath, CustomRule customRule, string? moniker =
 
             return customRule.ContentTypes is null || customRule.ContentTypes.Contains(pageType);
         }
+
+        public Error WithCustomRule(Error error)
+        {
+            if (TryGetCustomRule(error, _customRules, out var customRule))
+            {
+                error = WithCustomRule(error, customRule);
+            }
+            return error;
+        }
+
+        public static Error WithCustomRule(Error error, CustomRule customRule, bool? enabled = null)
+        {
+            var level = customRule.Severity ?? error.Level;
+
+            if (level != ErrorLevel.Off && customRule.ExcludeMatches(error.OriginalPath ?? error.Source?.File?.Path ?? ""))
+            {
+                level = ErrorLevel.Off;
+            }
+
+            if (enabled != null && !enabled.Value)
+            {
+                level = ErrorLevel.Off;
+            }
+
+            var message = error.Message;
+
+            if (!string.IsNullOrEmpty(customRule.Message))
+            {
+                try
+                {
+                    message = string.Format(customRule.Message, error.MessageArguments);
+                }
+                catch (FormatException)
+                {
+                    message += "ERROR: custom message format is invalid, e.g., too many parameters {n}.";
+                }
+            }
+
+            message = string.IsNullOrEmpty(customRule.AdditionalMessage) ?
+                message : $"{message}{(message.EndsWith('.') ? "" : ".")} {customRule.AdditionalMessage}";
+
+            return new Error(
+                level,
+                string.IsNullOrEmpty(customRule.Code) ? error.Code : customRule.Code,
+                message,
+                error.MessageArguments,
+                error.Source,
+                error.PropertyPath,
+                error.OriginalPath,
+                customRule.PullRequestOnly);
+        }
+
+        private bool TryGetCustomRule(
+            Error error,
+            Dictionary<string, List<SourceInfo<CustomRule>>> allCustomRules,
+            out SourceInfo<CustomRule>? customRule)
+        {
+            if (allCustomRules.TryGetValue(error.Code, out var customRules))
+            {
+                foreach (var rule in customRules)
+                {
+                    // todo need confirm override order
+                    var r = rule.Value;
+                    if (r.PropertyPath != null)
+                    {
+                        // compare with code + propertyPath + contentType
+                        var source = error.Source?.File;
+                        var pageType = source != null ? _documentProvider.GetPageType(source) : null;
+                        if (r.PropertyPath.Equals(error.PropertyPath) && r.ContentTypes.Contains(pageType))
+                        {
+                            customRule = rule;
+                            return true;
+                        }
+                        continue;
```

Comment:
>continue [](start = 24, length = 8)

useless?

Reply:
correct, it is unreachable in current code.